GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs

```diff
@@ -0,0 +1,123 @@
+﻿using GVFS.Common.Git;
+using GVFS.Common.Tracing;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+
+namespace GVFS.Common.Maintenance
+{
+    public class LooseObjectsStep : GitMaintenanceStep
+    {
+        private const string LooseObjects = "loose-objects.history";
+        private readonly string looseObjectsHistoryPath = null;
+        private readonly bool ignoreTimeRestriction = false;
+
+        public LooseObjectsStep(GVFSContext context, bool requireCacheLock, bool ignoreTimeRestriction)
+            : this(context, requireCacheLock)
+        {
+            this.ignoreTimeRestriction = ignoreTimeRestriction;
+        }
+
+        public LooseObjectsStep(GVFSContext context, bool requireCacheLock)
+            : base(context, gitObjects: null, requireObjectCacheLock: requireCacheLock)
+        {
+            this.looseObjectsHistoryPath = Path.Combine(this.Context.Enlistment.GitObjectsRoot, "info", LooseObjects);         
+        }
+
+        public override string Area => "LooseObjectsStep";
+
+        protected override bool PerformMaintenance()
+        {
+            using (ITracer activity = this.Context.Tracer.StartActivity("LooseObjectsStep", EventLevel.Informational, Keywords.Telemetry, metadata: null))
+            {
+                try
+                {
+                    if (!this.ignoreTimeRestriction && !this.EnoughTimeBetweenRuns())
+                    {
+                        activity.RelatedWarning("Skipping step due to not enough time between runs");
+                        return false;
+                    }
+
+                    // Verify no git processes are running
+                    IEnumerable<int> processIds = this.RunningGitProcessIds();
+                    if (processIds.Count() > 0)
+                    {                            
+                        activity.RelatedWarning("Skipping LooseObjectsStep due to git pids {0}", string.Join(",", processIds));
+                        return false;
+                    }
+
+                    // Perform prune-packed

```

Comment:
To confirm, is the goal to prune loose objects that were created locally? (i.e. not those that were downloaded to the shared cache?)

Reply:
It should use the shared cache (i.e.) D:\.gvfsCache\a34d3fab679e49d191e525a2f1ba1\gitObjects

Reply:
our goal is to not touch their `src/.git` folder (at this time). Instead, we will delete loose objects in their shared object cache _that already exist in packfiles in the shared object cache_.

Reply:
So one test that is planned is to run the step and verify that only known duplicate objects are deleted and other loose objects remain.

Reply:
It can be helpful to point to #488 for the 4-step plan for shared object cache maintenance.

Reply:
Do we need to point `prune-packed` at the shared cache?  It looks like it's running against `src/.git`:

```
        public Result PrunePacked()
        {
            return this.InvokeGitAgainstDotGitFolder("prune-packed -q");
        }
```

Reply:
@wilbaker this is a weird one because it uses an environment variable to find the scared cache.  
https://git-scm.com/docs/git-prune

It also seems inconsistent with the way the object dir is sent to git.  Maybe it does make sense for this to take in a parameter, and then pass that down to InvokeGitAgainstDogGitFolder to set in the environment.

Reply:
Thanks for the details here.  Does `prune-packed` care about the .git folder at all?

I'm wondering if we should be calling `InvokeGitAgainstDotGitFolder` at all because it's a bit weird that we're passing in `--git-dir=` if git just ends up ignoring it.

Should we just call `InvokeGitImpl` directly in `PrunePacked()`?

Reply:
It's worth testing to see if `prune-packed` doesn't need a `.git` dir, but it may complain "I'm not in a git repo"